### PR TITLE
[WIP] Data pipeline and DynamoDB backup management

### PR DESCRIPTION
**Pull request by bwarfield-amplify**
New code (library and script entry points) for creating and restoring backups for DynamoDB tables, including not incidentally a bunch of code for managing data pipelines.

NOTE: synthetic PR for early review purposes! This PR is against a temp branch that includes branches for two pending PRs (#175 and #155), because there's a reason those PRs are out there and it's not because I didn't want to use that code in upcoming new development. 😁 

TODO (aka DONTBUGMEABOUT):

  * [ ] unit tests for disco_datapipeline package
  * [x] disentangle DP management from DDB backup management

### CODE REVIEW

File: disco_aws_automation/disco_datapipeline.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
+"""
+Back end behaviors for creating data pipelines.
```
Comment (bwarfield-amplify) [inline disco_aws_automation/disco_datapipeline.py]: What's this? An entire new file with no unit tests?  Say it ain't so!

Yeah, it's a TODO.

File: disco_aws_automation/disco_datapipeline.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+                    return False
+            return True
+
+        for i in range(0, len(id_objects), window):
+            batch = throttled_call(self._dp_client.describe_pipelines,
+                                   pipelineIds=[desc['id'] for desc in id_objects[i:i + window]])
+            descriptions.extend([desc for desc in batch['pipelineDescriptionList'] if _search_matches(desc)])
+        return [
+            AsiaqDataPipeline(
+                pipeline_id=meta['pipelineId'], name=meta['name'], description=meta.get('description'),
```
Comment (bsuzukiamplify) [inline disco_aws_automation/disco_datapipeline.py]: I'm missing something, where is `meta` coming from?

File: disco_aws_automation/disco_datapipeline.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+    def fetch_all_descriptions(self):
+        """
+        Fetch all pipelines in this account/region, populating only their name/metadata fields
+        (use fetch_content to get the low-level details).
+        """
+        return self.search_descriptions()
+
+    def search_descriptions(self, name=None, tags=None):
+        "Fetch all pipelines in this account/region that have the given tags and/or the given name."
```
Comment (bsuzukiamplify) [inline disco_aws_automation/disco_datapipeline.py]: Should we be able to search by pipeline id as well?

File: disco_aws_automation/asiaq_cli.py [hunk 4]
```diff
@@ -103,7 +104,125 @@ def _update_s3_configs(self, sandbox_name):
...
+                                    help="Previous backup to restore from (default: latest)")
+        list_parser = subsub.add_parser("list", help="List available pipelines")
+        list_parser.add_argument("--pipeline-name", dest="search_name", help="Find pipelines with this name.")
+        list_parser.add_argument("--all-envs", dest="ignore_env", action='store_const', const=True)
+
+    def run(self):
+        dispatch = {
+            'list': self._search,
+            'configure_dynamodb_backup': self._create_backup,
+            'restore_dynamodb': self._restore_backup,
```
Comment (bsuzukiamplify) [inline disco_aws_automation/asiaq_cli.py]: Actually I see support for delete in the code below so presumably just not added at this level yet.  Cool.

File: disco_aws_automation/disco_datapipeline.py (hunk 2 above)
Comment (bwarfield-amplify) [inline disco_aws_automation/disco_datapipeline.py]: Line 140—is a list comprehension. 😸

File: disco_aws_automation/disco_datapipeline.py (hunk 3 above)
Comment (bwarfield-amplify) [inline disco_aws_automation/disco_datapipeline.py]: That's just `fetch`: you can have multiple pipelines with the same name/tags, but the ID is unique.

**Comment by bwarfield-amplify** [comment]
Replaced by #182

**Pull request closed without merging**